src/Analysis/Ast/Impl/Diagnostics/DiagnosticsEntry.cs

```diff
@@ -51,6 +53,20 @@ public sealed class DiagnosticsEntry {
         /// </summary>
         public DiagnosticSource Source { get; }
 
+        public bool ShouldReport(IPythonModule module) {
+            // module should always be a user written python module
+            if (module.ModuleType != ModuleType.User || !(module is PythonModule pythonModule)) {
+                return false;
+            }
+
+            string line = pythonModule.GetLine(SourceSpan.Start.Line);
+            if (line.Contains("#noqa")) {
```

Comment:
This is too simple; `# noqa` is also valid, and `# noqa: E123` should only disable whatever's associated with that code.